src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,134 @@
+        }
+    }
+
+    private Object[] grow() {
+        return grow(size + 1);
+    }
+
+    private Object[] grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0 || elementData != EMPTY_ELEMENT_DATA) {
+            int newCapacity = oldCapacity >> 1;
+            Object[] copyOfElementData = elementData;
+            elementData = new Object[Math.max(newCapacity,minCapacity)];
+            System.arraycopy(copyOfElementData,0, elementData, 0, oldCapacity);
+            return elementData;
+        }
+        return elementData = new Object[Math.max(DEFAULT_CAPACITY, minCapacity)];
+    }
+
+    private void add(T e,Object[] element,int size) {
+        if (size == element.length) {
+            element = grow();
+        }
+        element[size] = e;
+        this.size = size + 1;
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, elementData, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        add(value, elementData, size);
+        checkIndex(index);
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] inputListToArray = new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            inputListToArray[i] = list.get(i);
+        }
+        if (inputListToArray.length > elementData.length - size) {
+            elementData = grow(size + inputListToArray.length);
+        }
+        System.arraycopy(inputListToArray, 0, elementData, size, inputListToArray.length);
+        size = size + inputListToArray.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
+        checkIndex(index);
+        elementData[index] = value;
 
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = (T) elementData[index];
+        quickRemove(elementData, index);
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = 0;
+        found: {
+            if (element == null) {
+                for (; index < size; index++) {
+                    if (elementData[index] == null) {
+                        break found;
+                    }
+                }
+            } else {
+                for (; index < size; index++) {
+                    if (element.equals(elementData[index])) {
+                        break found;
+                    }
+                }
+            }
+            throw new NoSuchElementException("element not found");
```

Comment:
The exception message 'element not found' is not very informative. It would be better to include the actual element that was not found in the message.